edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/routing/CloudEndpoint.cs

```diff
@@ -175,6 +123,75 @@ public Task CloseAsync(CancellationToken token)
+                var sendFailureDetails = new SendFailureDetails(FailureKind.Transient, new EdgeHubConnectionException("IoT Hub is not connected"));
+
+                return Task.FromResult((ISinkResult<IRoutingMessage>)new SinkResult<IRoutingMessage>(ImmutableList<IRoutingMessage>.Empty, failed, sendFailureDetails));
+            }
+
+            async Task<ISinkResult<IRoutingMessage>> ProcessAsync(IRoutingMessage routingMessage, string identity)
+            {
+                IMessage message = this.cloudEndpoint.messageConverter.ToMessage(routingMessage);
+                Util.Option<ICloudProxy> cloudProxy = await this.cloudEndpoint.cloudProxyGetterFunc(identity);
+
+                ISinkResult result = await cloudProxy.Match(
+                    async cp =>
+                    {
+                        try
+                        {
+                            using (Metrics.CloudLatency(identity))
+                            {
+                                await cp.SendMessageAsync(message);
+                            }
+
+                            var succeeded = new List<IRoutingMessage> { routingMessage };
+                            Metrics.MessageCount(identity);
+
+                            return new SinkResult<IRoutingMessage>(succeeded);
+                        }
+                        catch (Exception ex)
+                        {
+                            return this.HandleException(ex, routingMessage);
+                        }
+                    },
+                    () => this.ProcessNoConnection(identity, routingMessage)
+                );
+
+                return result;
+            }
+
+            ISinkResult HandleException(Exception ex, IRoutingMessage routingMessage)
+            {
+                if (IsRetryable(ex))
+                {
+                    var failed = new List<IRoutingMessage> { routingMessage };
+
+                    Events.RetryingMessage(routingMessage, ex);
+                    var sendFailureDetails = new SendFailureDetails(FailureKind.Transient, new EdgeHubIOException($"Error sending messages to IotHub for device {this.cloudEndpoint.Id}"));
+
+                    return new SinkResult<IRoutingMessage>(ImmutableList<IRoutingMessage>.Empty, failed, sendFailureDetails);
+                }
+                else
+                {
+                    Events.InvalidMessage(ex);
+                    var invalid = new List<InvalidDetails<IRoutingMessage>> { new InvalidDetails<IRoutingMessage>(routingMessage, FailureKind.InvalidInput) };
+                    var sendFailureDetails = new SendFailureDetails(FailureKind.InvalidInput, ex);
+
+                    return new SinkResult<IRoutingMessage>(ImmutableList<IRoutingMessage>.Empty, ImmutableList<IRoutingMessage>.Empty, invalid, sendFailureDetails);
+                }
+            }
+
+            Task<ISinkResult> ProcessNoIdentity(IRoutingMessage routingMessage)
```

Comment:
Should be "ProcessNoIdentityAsync" since it returns Task<>.

Reply:
We don't really follow this convention everywhere in our code.